### PR TITLE
test: avoid noise in test logs

### PR DESCRIPTION
**Pull request by justinmk**
# Problem:

Tests that _intentionally_ fail certain conditions cause noise in $NVIM_LOG_FILE:

    $NVIM_LOG_FILE: /home/runner/work/neovim/neovim/build/.nvimlog
    (last 100 lines)
    WRN 2023-01-16T18:26:27.673 T599.7799.0 unsubscribe:519: RPC: ch 1: tried to unsubscribe unknown event 'doesnotexist'
    WRN 2023-01-16T18:29:00.557 ?.11151    server_start:163: Failed to start server: no such file or directory: /X/X/X/...
    WRN 2023-01-16T18:33:07.269 127.0.0.1:12345 server_start:163: Failed to start server: address already in use: 127.0.0.1
    ...
    -- Output to stderr:
    module 'vim.shared' not found:
    	no field package.preload['vim.shared']
    	no file './vim/shared.lua'
    	no file '/home/runner/nvim-deps/usr/share/lua/5.1/vim/shared.lua'
    	no file '/home/runner/nvim-deps/usr/share/lua/5.1/vim/shared/init.lua'
    	no file '/home/runner/nvim-deps/usr/lib/lua/5.1/vim/shared.lua'
    	no file '/home/runner/nvim-deps/usr/lib/lua/5.1/vim/shared/init.lua'
    	no file './vim/shared.so'
    	...
    E970: Failed to initialize builtin lua modules

# Solution:

- Log to a private $NVIM_LOG_FILE in tests that intentionally fail and cause ERR log messages.
- Assert that the expected messages are actually logged.

### CODE REVIEW

File: test/functional/options/defaults_spec.lua [hunk 1]
```diff
@@ -422,6 +436,9 @@ describe('XDG defaults', function()
     end)
 
     it('are not expanded', function()
+      retry(nil, 1000, function()
+        assert_log('Failed to start server: no such file or directory: %$XDG_RUNTIME_DIR%/', testlog, 10)
```
Comment (justinmk) [inline test/functional/options/defaults_spec.lua:440]: `assert_log` should probably do `retry` implicitly, it's always needed...

File: test/functional/lua/vim_spec.lua [hunk 2]
```diff
@@ -2910,9 +2910,14 @@ describe('lua: builtin modules', function()
   end)
 
 
-  it('does not work when disabled without runtime', function()
-    clear{args={'--luamod-dev'}, env={VIMRUNTIME='fixtures/a'}}
-    expect_exit(exec_lua, [[return vim.tbl_count {x=1,y=2}]])
+  it('fails when disabled without runtime', function()
+    clear()
+    command("let $VIMRUNTIME='fixtures/a'")
```
Comment (justinmk) [inline test/functional/lua/vim_spec.lua:2915]: fails correctly when this line is removed

**Comment by zeertzjq** [comment]
CI failure

**Comment by justinmk** [comment]
fixing

**Comment by justinmk** [comment]
fixed https://github.com/neovim/neovim/commit/6c2f02cbd02e8fd90e8ffae271dad3df48a54c4a

**Pull request merged**